File: cli/raft-tools/libs/node-js/raft.js
'use strict';

const http = require('http');
const querystring = require('querystring');
const fs = require('fs');
const path = require('path'); 
const { exec } = require('child_process');

const workDirectory = process.env.RAFT_WORK_DIRECTORY;
const agentUtilitiesUrl = process.env.RAFT_AGENT_UTILITIES_URL;
console.log(`AgentUtilitiesURL: ${agentUtilitiesUrl}`);

const rawdata = fs.readFileSync(workDirectory + '/task-config.json');
const config = JSON.parse(rawdata);

function jsonGet(jo, keys) {
    var d = jo;
    for (const k of keys) {
        d = d[Object.keys(d).find(key => k.toLowerCase() === key.toLowerCase())];
        if (d == null) {
            break;
        }
    }
    return d;
}

const timer = ms => new Promise(res => setTimeout(res, ms));
function waitForEndpoint(url, callback) {
    http.get(url,
        function(res) {
            if (res.statusCode !== 200) {
                timer(1000).then( _ => waitForEndpoint(url, callback));
            } else {
                callback(null, null);
            }
        }
    ).on("error", function(err) {
            console.log(`Failed to establish connection to ${url} due to ${err}. Trying again...`);
            timer(1000).then( _ => waitForEndpoint(url, callback));
        }
    )
}



class RaftUtils {
    constructor(toolName) {
        this.jobId = process.env['RAFT_JOB_ID'];
        this.agentName = process.env['RAFT_CONTAINER_NAME'];
        this.traceProperties =
            {
                jobId : this.jobId,
                taskIndex : process.env['RAFT_TAKS_INDEX'],
                containerName : this.agentName
            };
        this.toolName = toolName;

        this.agentUtilitiesUrl = new URL(agentUtilitiesUrl);
    }

    post(path, data) {
        var postData = JSON.stringify(data);
        const opts = {
            host : this.agentUtilitiesUrl.hostname,
            port : this.agentUtilitiesUrl.port,
            path : path,
            method : 'POST',
            headers : {
                'Content-Type': 'application/json',
                'Content-Length' : Buffer.byteLength(postData)
            }
        };

        return new Promise(function(resolve, reject) {
                var post = http.request(opts, res => {
                    res.on('data', d => {
                        let rawData = '';
                        res.on('data', (chunk) => { rawData += chunk; });
                        res.on('end', () => {
                            try {
                                const parsedData = JSON.parse(rawData);
                                resolve(parsedData['token']);
                            } catch (e) {
                                reject(e);
                            }
                        }).on('error', (e) => {
                            reject(e);
                        });
                    });
                });

                post.on('error', e => {
                    console.error(`Posting ${postData} to ${path} failed with ${e}`);
                    reject(e);
                });

                post.write(postData);
                post.end();
            }
        );
    }

    waitForAgentUtilities(callback) {
        const readinessUrl = agentUtilitiesUrl + '/readiness/ready';
        return waitForEndpoint(readinessUrl, callback);
    }

    postEvent (eventName, eventData) {
        return this.post(('/messaging/event/' + eventName), eventData);
    }

    reportBug(bugDetails) {
        var bugData = {
            tool : this.toolName,
            jobId : this.jobId,
            agentName : this.agentName,
            bugDetails : bugDetails
        };
        return this.postEvent('bugFound', bugData);
    }

    reportStatus(state, details) {
        var jobStatusData = {
            tool: this.toolName,
            jobId : this.jobId,
            agentName : this.agentName,
            details : details,
            utcEventTime : (new Date()).toUTCString(),
            state : state
        };
        return this.postEvent('jobStatus', jobStatusData);
    }

    reportStatusCreated(details){
        return this.reportStatus('Created', details);
    }

    reportStatusRunning(details){
        return this.reportStatus('Running', details);
    }

    reportStatusCompleted(details){
        return this.reportStatus('Completed', details);
    }

    reportStatusError(details){
        return this.reportStatus('Error', details);
    }

    logTrace(traceMessage) {
        return this.post('/messaging/trace', {message: traceMessage, tags : this.traceProperties, severity : 'Information'});
    }

    logException(exception) {
        return this.post('/messaging/trace', {message: exception, tags : this.traceProperties, severity : 'Error'});
    }

    flush(){
        return this.post('/messaging/flush');
    }
}
function getAuthHeader(callback) {
    const authenticationMethod = jsonGet(config, ['authenticationMethod']);
    if (!authenticationMethod)  {
        callback(null, null);
    } else {
        const authMethods = Object.keys(authenticationMethod);
        if (authMethods.length == 0) {
            callback(null, null); 
        } else if (authMethods.length > 1) {
            callback(new Error("More than one authentication method is specified: " + authenticationMethod), null);
        } else {
            const authMethod = authMethods[0];
            console.log("Authentication Method: " + authMethod);
            const authUrl = agentUtilitiesUrl + `/auth/${authMethod}/${authenticationMethod[authMethod]}`
            http.get(authUrl,
                function(res) {
                    if (res.statusCode !== 200) {
                        callback(new Error(`Request to ${authUrl} failed with status code ${res.statusCode}`), null);
                    } else {
                        let rawData = '';
                        res.on('data', (chunk) => { rawData += chunk; });
                        res.on('end', () => {
                            try {
                                const parsedData = JSON.parse(rawData);
                                callback(null, parsedData['token']);
                            } catch (e) {
                                callback(e, null);
                            }
                        }).on('error', (e) => {
                            callback(e, null);
                        });
                    }
                }
            );
        }
    }
}

function installCertificates(callback) {
    const certificates = jsonGet(config, ['targetConfiguration', 'certificates'])

    if (!certificates) {
        callback(null, null);
    } else {
        fs.readdir(certificates, function(err, files) {
            if (err) {
                callback(err, null);
            }
            else {
                files.forEach(function(file) {
                    console.log("File: " + file);
                    if (path.extname(file) === '.crt') {
                        const copySrc = certificates + "/" + file;
                        const copyDest = "/usr/local/share/ca-certificates/" + file;
                        console.log("CopySrc: " + copySrc + " CopyDest: " + copyDest);
                        fs.copyFileSync(copySrc, copyDest);
                    }
                });
                console.log("Updating certificates");
                exec("update-ca-certificates --fresh", (error, _) => {
                        if (error) {
                            console.log("Failed to update certificates: " + error);
                            callback(error, null);
                        } else {
                            callback(null, null);
                        }
                    }
                );
            }
        });
    }
}

exports.installCertificates = installCertificates;
exports.workDirectory = workDirectory;
exports.config = config;
exports.RaftUtils = RaftUtils;
exports.getAuthHeader = getAuthHeader;
exports.jsonGet = jsonGet;

File: cli/raft-tools/libs/python3/raft.py
import json
import os
import subprocess
import sys
import time
import io
import importlib
import shutil
import glob
import requests

from urllib.parse import urlparse
from contextlib import redirect_stdout

class RaftJsonDict(dict):
    def __init__(self):
        super(RaftJsonDict, self).__init__()

    def __getitem__(self, key):
        for k in self.keys():
            if k.lower() == key.lower():
                key = k
                break
        return super(RaftJsonDict, self).__getitem__(key)

    def pop(self, key):
        for k in self.keys():
            if k.lower() == key.lower():
                key = k
                break
        return super(RaftJsonDict, self).pop(key)

    def get(self, key):
        for k in self.keys():
            if k.lower() == key.lower():
                key = k
                break
        return super(RaftJsonDict, self).get(key)

    @staticmethod
    def raft_json_object_hook(x):
        r = RaftJsonDict()
        for k in x:
            r[k] = x[k]
        return r

def install_certificates():
    work_directory = os.environ['RAFT_WORK_DIRECTORY']
    run_directory = os.environ['RAFT_TOOL_RUN_DIRECTORY']
    with open(os.path.join(work_directory, "task-config.json"), 'r') as task_config:
        config = json.load(task_config, object_hook=RaftJsonDict.raft_json_object_hook)
        if config.get("targetConfiguration") and config.get("targetConfiguration").get("certificates"):
            certificates = config["targetConfiguration"]["certificates"]
            files = glob.iglob(os.path.join(certificates, "*.crt"))
            for f in files:
                if os.path.isfile(f):
                    print(f"Copying file {f}")
                    shutil.copy(f, "/usr/local/share/ca-certificates/")
            subprocess.check_call(["update-ca-certificates", "--fresh"])


def auth_token():
    auth_url = os.environ['RAFT_AGENT_UTILITIES_URL'] 
    work_directory = os.environ['RAFT_WORK_DIRECTORY']
    run_directory = os.environ['RAFT_TOOL_RUN_DIRECTORY']
    with open(os.path.join(work_directory, "task-config.json"), 'r') as task_config:
        config = json.load(task_config, object_hook=RaftJsonDict.raft_json_object_hook)
        auth_config = config.get("authenticationMethod")
        if auth_config:
            auth_type, auth_key = auth_config.popitem()
            response = requests.get(auth_url + '/auth' + '/' + auth_type + '/' + auth_key)
            if response.ok:
                content = json.loads(response.text)
                return content['token']
            else:
                raise Exception(response.text)
        else:
            return None


def task_config():
    work_directory = os.environ['RAFT_WORK_DIRECTORY']
    with open(os.path.join(work_directory, 'task-config.json'), 'r') as task_config:
        return json.load(task_config, object_hook=RaftJsonDict.raft_json_object_hook)

class RaftUtils():
    def __init__(self, tool_name):
        self.config = task_config()
        self.report_status_url = os.environ['RAFT_AGENT_UTILITIES_URL']
        self.job_id = os.environ['RAFT_JOB_ID']
        self.container_name = os.environ['RAFT_CONTAINER_NAME']
        self.tool_name = tool_name

        self.telemetry_properties = {
            "jobId" : self.job_id,
            "taskIndex" : os.environ['RAFT_TASK_INDEX'],
            "containerName" : self.container_name
        }

    def wait_for_agent_utilities(self):
        try:
            r = requests.get(f'{self.report_status_url}/readiness/ready')
            if r.ok:
                return
            else:
                time.sleep(1)
                self.wait_for_agent_utilities()
        except:
            time.sleep(10)
            return self.wait_for_agent_utilities()

    def report_bug(self, bugDetails):
        m = {
            'tool' : self.tool_name,
            'jobId' : self.job_id,
            'agentName' : self.container_name,
            'bugDetails' : bugDetails
        }
        requests.post(f'{self.report_status_url}/messaging/event/bugFound', json=m)

    def report_status(self, state, details):
        m = {
            'tool' : self.tool_name,
            'jobId' : self.job_id,
            'agentName': self.container_name,
            'details': details,
            'utcEventTime' : time.strftime('%Y-%m-%d %H:%M:%S', time.gmtime()),
            'state' : state
        }
        requests.post(f'{self.report_status_url}/messaging/event/jobStatus', json=m)

    def report_status_created(self, details=None):
        self.report_status('Created', details)

    def report_status_running(self, details=None):
        self.report_status('Running', details)

    def report_status_error(self, details=None):
        self.report_status('Error', details)

    def report_status_completed(self, details=None):
        self.report_status('Completed', details)

    def log_trace(self, trace):
        t = {
            'message' : trace,
            'severity' : 'Information',
            'tags' : self.telemetry_properties
        }
        requests.post(f'{self.report_status_url}/messaging/trace', json=t)

    def log_exception(self, ex):
        t = {
            'message' : f'{ex}',
            'severity' : 'Error',
            'tags' : self.telemetry_properties
        }
        requests.post(f'{self.report_status_url}/messaging/trace', json=t)

    def flush(self):
        requests.post(f'{self.report_status_url}/messaging/flush')
